File: src/core/Mailbox.ts
import Mail from "nodemailer/lib/mailer"
import { MailboxName, Email, UniqueIdentifier } from "."

export class Mailbox {
  static named(name: string | MailboxName) {
    return new this(UniqueIdentifier.unknown(), MailboxName.of(name), [])
  }

  private constructor(
    public readonly id: UniqueIdentifier,
    public readonly name: MailboxName,
    public readonly emails: Email[]
  ) {}

  public withId(id: UniqueIdentifier) {
    return new Mailbox(id, this.name, this.emails)
  }

  public withEmails(emails: Email[]) {
    return new Mailbox(this.id, this.name, emails)
  }
}

File: inbox_coach/lib/inbox_coach/mailbox.ex
defmodule InboxCoach.Mailbox do
  defstruct [:id, :name]

  def new(mailbox) do
    %__MODULE__{
      id: mailbox["id"],
      name: mailbox["name"]
    }
  end

  def name(mailbox) do
    mailbox["name"]
  end
end
